lib/wallaroo/core/source/connector_source/connector_framed_source_notify.pony

```diff
@@ -82,91 +254,348 @@ class ConnectorSourceNotify[In: Any val]
+              ifdef "resilience" then
+                _pending_close.update(m.stream_id, s)
+              else
+                // respond immediately
+                _send_reply(source, cwm.AckMsg(0, [(s.id, s.last_seen)]))
+                _listener.streams_relinquish([StreamTuple(s.id, s.name,
+                  s.last_seen)])
+                // TODO [source-migration] remove stream from active
+              end
+              return true
+            elseif cwm.Boundary.is_set(m.flags) then
+              // TODO [post-source-migration] what's supposed to happen here?
+              return true
+            else // not an EOS and not a boundary
+              // process message
+              try
+                // get bytes content of message
+                let bytes = match (m.message as cwm.MessageBytes)
+                | let str: String      => str.array()
+                | let b: Array[U8] val => b
+                end
+
+                // decode bytes using handler, if bytes not None
+                let decoded = if bytes.size() == 0 then
+                  None
+                else
+                  _handler.decode(bytes)?
+                end
+
+                ifdef "trace" then
+                  @printf[I32](("Msg decoded at " + _pipeline_name +
+                    " source\n").cstring())
+                end
+
+                // get message key
+                let key_string =
+                  match m.key
+                  | None =>
+                    ""
+                  | let str: String val =>
+                    str
+                  | let a: Array[U8] val =>
+                    let k' = recover trn String(a.size()) end
+                    for c in a.values() do
+                      k'.push(c)
+                    end
+                    consume k'
+                  end
+
+                // process message
+                return _run_and_subsequent_activity(latest_metrics_id, ingest_ts,
+                  pipeline_time_spent, key_string, source, decoded, s,
+                  m.message_id, m.flags)
+              else
+                // _handler.decode(bytes) failed
+                if m.message is None then
+                  // TODO [post-source-migration] revisit this error message
+                  return _to_error_state(source, "No message bytes and BOUNDARY not set")
+                end
+                @printf[I32](("Unable to decode message at " + _pipeline_name + " source\n").cstring())
+                ifdef debug then
+                  Fail()
+                end
+                return _to_error_state(source, "Unable to decode message")
+              end
+            end
+          else
+            return _to_error_state(source, "Unknown StreamId")
           end
-          (true, ingest_ts)
         end
 
-      if is_finished then
-        let end_ts = WallClock.nanoseconds()
-        let time_spent = end_ts - ingest_ts
+      | let m: cwm.AckMsg =>
```

Comment:
Are these invalid messages ones that we intend to add to the protocol soon?  If not, can we move them all into a single method that checks for invalid messages so they don't pollute this match?

Reply:
These will remain invalid messages, I believe, according to `connector-protocol-v2.md`.  Moving them elsewhere is fine.